dm/syncer/binlogstream/streamer_controller.go

```diff
@@ -354,28 +326,56 @@ func (c *StreamerController) GetEvent(tctx *tcontext.Context) (
 					return nil, 0, pb.ErrorOp_InvalidErrorOp, terror.Annotate(err2, "fail to parse binlog file name from rotate event")
 				}
 				ev.NextLogName = []byte(utils.ConstructFilenameWithUUIDSuffix(filename, c.relaySubDirSuffix))
-				event.Event = ev
 			}
 		}
-		return event, suffix, op, nil
 	}
 
-	checkErr := func(err error) (shouldRet bool) {
-		if err == nil {
-			return false
+	if err != nil {
+		if err == context.Canceled || err == context.DeadlineExceeded {
+			return nil, 0, pb.ErrorOp_InvalidErrorOp, err
 		}
+		c.Lock()
+		tctx.L().Error("meet error when get binlog event", zap.Error(err))
+		c.meetError = true
+		c.Unlock()
+		return nil, 0, pb.ErrorOp_InvalidErrorOp, err
+	}
 
-		if err != context.Canceled && err != context.DeadlineExceeded {
-			c.Lock()
-			tctx.L().Error("meet error when get binlog event", zap.Error(err))
-			c.meetError = true
-			c.Unlock()
+	if suffix != 0 {
+		c.locations.curStartLocation = c.upstream.locationRecorder.curStartLocation
+		c.locations.curStartLocation.Suffix = suffix - 1
+		c.locations.curEndLocation = c.upstream.locationRecorder.curStartLocation
+		c.locations.curEndLocation.Suffix = suffix
+		c.locations.txnEndLocation = c.locations.curEndLocation
+	} else {
+		if c.locations.curEndLocation.Suffix != 0 {
+			// this is the first upstream event after injection, keep suffix for start location
+			c.locations.curStartLocation = c.locations.curEndLocation
+			c.locations.curEndLocation = c.upstream.locationRecorder.curEndLocation
+			c.locations.txnEndLocation = c.upstream.locationRecorder.txnEndLocation
+		} else {
+			c.locations = c.upstream.locationRecorder.locations
 		}
-		return true
 	}
 
+	return event, suffix, op, nil
+}
+
+func (c *StreamerController) getEvent(tctx *tcontext.Context) (
+	event *replication.BinlogEvent,
+	suffix int,
+	op pb.ErrorOp,
+	err error,
+) {
+	failpoint.Inject("SyncerGetEventError", func(_ failpoint.Value) {
+		c.meetError = true
+		tctx.L().Info("mock upstream instance restart", zap.String("failpoint", "SyncerGetEventError"))
+		failpoint.Return(nil, 0, pb.ErrorOp_InvalidErrorOp, terror.ErrDBBadConn.Generate())
+	})
+
+	// TODO: why this lock?
 	c.RLock()
-	streamer := c.streamer
+	upstream := c.upstream
```

Comment:
Are there any risks if we use `c.upstream` directly (e.g. will `getEvent` be invoked concurrently while the `c.upstream` is changed by other threads)?

Reply:
seems no, will remove it